In the first notebook [notebook_01_webscraping_Evanescence_Within_Temptation.ipynb](http://localhost:8888/notebooks/Project_Evanescence_Within_Temptation/notebook_01_webscraping_Evanescence_Within_Temptation.ipynb) of this project we used web scrapping to obtain lyrics of `Evanescence` and `Within Temptation`. After that we used Spotify API to retrieve more details about both bands including information about them, their albums, and details about their lyrics including not only metadata (e.g. ) but also audio features (e.g. valence, energy, tempo, liveness) [notebook_02_retrieve_Spotify_data-Evanescence_Within_Temptation.ipynb](http://localhost:8888/notebooks/Project_Evanescence_Within_Temptation/notebook_02_retrieve_Spotify_data-Evanescence_Within_Temptation.ipynb).

Now is time to use the data retrieved and try to explore and visualize as much as we can. Our goal is not only explore text data, but also visualize numeric and categorical features. 

In what concerns NLP (Natural Language Processing) I want to do some things:

1. Text analysis: Analyze both bands and compare them through their lyrics using some metrics and word clouds.
2. Sentiment analysis: Explore the sentiment, polarity, and subjectivity of the lyrics provided by [TextBlob](https://textblob.readthedocs.io/en/dev/index.html) to compare both bands through visualization.
3. I'll try also to connect metadata of tracks with the sentiment provided by lyrics to draw conclusions.
4. Analyse some of the audio features, in special the ones that have been pointed as mood features, i.e., valence and energy, and see if there is a relation between them and the sentiment of lyrics of a track.

Let’s get started!

# Loading all data

## Lyrics

In [1]:
import pandas as pd


In [2]:
df_lyrics_evanescence = pd.read_csv("./data/lyrics_evanescence_2020-02-16.csv")
df_lyrics_evanescence.sort_values(by='song_title').head(20)

song_title  \
3                4th of july   
69    all that im living for   
50             angel of mine   
41          anything for you   
61                  anywhere   
39              away from me   
78           before the dawn   
59                     bleed   
67           breathe no more   
6           bring me to life   
12                    broken   
13  call me when youre sober   
9                 cloud nine   
72                 disappear   
8           end of the dream   
47                erase this   
25             even in death   
29           everybodys fool   
31                    exodus   
36             fall into you   

                                               lyrics  
3   Shower in the dark day. Clean sparks driving d...  
69  All that I'm living for. All that I'm dying fo...  
50  You are everything I need to see. Smile and su...  
41  I'd give anything to give me to you. Can you f...  
61  Dear my love, haven't you wanted to be with me...  
39  I hold my breath. as this life starts to take ...  
78  Meet me after dark again. and I'll hold you. I...  
59  How can I pretend that I don't see. What you h...  
67  I've been looking in the mirror for so long.. ...  
6   how can you see into my eyes. like open doors....  
12  I wanted you to know I love the way you laugh....  
13  Don't cry to me. If you loved me. You would be...  
9   If you want to live, let live.. If you want to...  
72  Hollow, like you don't remember me. Underneath...  
8   I found a grave. Brushed off the face. Felt yo...  
47  It's too late to change your mind. Even though...  
25  Give me a reason to believe that you're gone. ...  
29  perfect by nature. icons of self indulgence. j...  
31  My black backpack stuffed with broken dreams. ...  
36  Since you've been gone i've been alone. feelin...

In [3]:
df_lyrics_within_temptation = pd.read_csv("./data/lyrics_within_temptation_2020-02-16.csv")
df_lyrics_within_temptation.sort_values(by='song_title').head(15)

song_title                                             lyrics
17   a dangerous mind  Cause something is not right. I follow the sig...
64      a demons fate  Ooh, ooh, ooh, ooh, ooh. Ooh, ooh, ooh, ooh, o...
21         all i need  I'm dying to catch my breath. Oh why don't I e...
60             angels  Sparkling angel I believed. You were my saviou...
40        another day  I know you are going away. I take my love into...
51           aquarius  I hear your whispers. Break the silence and it...
29        bittersweet  If I tell you. Will you listen?. Will you stay...
67          blue eyes  Blue eyes wide to the world. Full of dreams an...
56              caged  These are the darkest clouds. They have surrou...
31            candles  Take away. These hands of darkness. Reaching f...
69         dark wings  Why was I one of the chosen ones?. Until the f...
47  deceiver of fools  He feeds on fear. He feeds on pain. He rules a...
63        deep within  Is there a place deep within. A place where yo...
27          destroyed  I did my best to please you. But my best was n...
43  edge of the world  Can I swallow the soil,. Make it all beautiful...

## Spotify's data

From all the data retrieved I'll concentrate on the track's information csv. I'll be using the one we saved in .csv that should have eliminated at least some duplicates from tracks.

In [4]:
df_tracks_evanescence = pd.read_csv("./data/info_tracks_evanescence_without_duplicates_2020-02-16.csv")
df_tracks_evanescence.sort_values(by='track_name').head(10)

album_name                track_id  \
51  Evanescence (Deluxe Version)  3UkDyGtriDY7NzOJbF0rIH   
37                 The Open Door  4iDQezFTnOwgnrPYiqQ6TP   
58                 Lost Whispers  2lH8hMXxuIcjpbIok9KbUj   
19      Anywhere But Home (Live)  2zn4moJkEmIVfV83iye9t5   
1                         Fallen  0COqiPhxzoWICwFCS4eZcp   
73                Synthesis Live  1rvxZ0qg96Nkr3PLhHTbCA   
21      Anywhere But Home (Live)  1AjCrY9w0edn2jAGEAkzJ7   
64                     Synthesis  4vHFFk4Vm9NWhGq2FAsTlj   
27                 The Open Door  663Karu2rvKLdnY0eo1n3M   
30                 The Open Door  6Sh05fnlrLbMfSuI8Qur6a   

                                           track_name  track_duration  \
51                                 a new way to bleed          226400   
37                           all that i am living for          228706   
58                   breathe no more   b side version          228809   
19  breathe no more   live from le zénith,france/2004          213853   
1                                    bring me to life          235893   
73                            bring me to life   live          264026   
21  bring me to life   live from le zénith,france/...          283760   
64                       bring me to life   synthesis          257320   
27                          call me when you're sober          214706   
30                                         cloud nine          262173   

    track_popularity                                      track_preview  \
51                44                                                NaN   
37                48                                                NaN   
58                49                                                NaN   
19                33                                                NaN   
1                 77                                                NaN   
73                29  https://p.scdn.co/mp3-preview/87cbd661e1853b8f...   
21                40                                                NaN   
64                 6                                                NaN   
27                64                                                NaN   
30                44                                                NaN   

    danceability  energy  key  loudness  mode  speechiness  acousticness  \
51         0.378   0.895    1    -4.347     1       0.0531       0.00005   
37         0.514   0.809    3    -4.396     0       0.0617       0.01210   
58         0.620   0.186   11    -8.527     0       0.0284       0.97100   
19         0.562   0.431   11   -10.670     0       0.0307       0.32300   
1          0.331   0.943    4    -3.188     0       0.0698       0.00721   
73         0.149   0.813    4    -5.260     0       0.0560       0.34600   
21         0.341   0.825    4    -7.220     0       0.0622       0.02210   
64         0.362   0.785    4    -3.876     0       0.0567       0.61000   
27         0.450   0.883    7    -4.094     1       0.0524       0.00193   
30         0.125   0.893    3    -4.217     0       0.2100       0.04320   

    instrumentalness  liveness  valence    tempo  
51          0.025200    0.1500   0.2580  155.946  
37          0.000000    0.0763   0.3850  136.881  
58          0.000001    0.1170   0.2190   96.992  
19          0.018500    0.9550   0.1670  108.012  
1           0.000002    0.2420   0.2960   94.612  
73          0.000021    0.9140   0.2420   90.642  
21          0.030600    0.5220   0.0398   94.992  
64          0.000001    0.0722   0.1600   90.904  
27          0.000000    0.2930   0.3280   93.410  
30          0.000085    0.1510   0.1900  194.550

In [5]:
df_tracks_within_temptation = pd.read_csv("./data/info_tracks_within_temptation_without_duplicates_2020-02-16.csv")
df_tracks_within_temptation.sort_values(by='track_name').head(10)

album_name                track_id  \
32                                 The Silent Force  6D5ih8y9mKmCSkuZO2Up2Q   
75                                  The Unforgiving  6ivwIJGFnzTRPG2dHvKA07   
42                          The Heart Of Everything  0lW4J9tzxpODQ8IExSumDW   
55                 An Acoustic Night At The Theatre  1tbSP6d2KwBB2DZUJLalRZ   
91                          Hydra (Special Edition)  6MubsJeQrVa0k7lJSxcdaM   
99                          Hydra (Special Edition)  13cZ2hORsadxvc2KLUBZoA   
131  Let Us Burn - Elements & Hydra Live In Concert  301osYEEEVs4EQNXZXStCi   
27                                 The Silent Force  3TEwbiC0GhIRStn3Eabtu7   
114  Let Us Burn - Elements & Hydra Live In Concert  6oQdvGElasxvHYutiewDSc   
105                               Enter + The Dance  4nroowkyOM1HB9BOwUVV3M   

                       track_name  track_duration  track_popularity  \
32               a dangerous mind          256533                34   
75                 a demon's fate          329537                40   
42                     all i need          290946                24   
55              all i need   live          320946                21   
91                     and we run          230067                 7   
99   and we run   evolution track          341497                 5   
131        and we run   live 2014          236746                 0   
27                         angels          240440                55   
114            angels   live 2012          252226                 0   
105                   another day          348453                16   

                                         track_preview  danceability  energy  \
32   https://p.scdn.co/mp3-preview/c2c47b037fe1394c...         0.365   0.894   
75   https://p.scdn.co/mp3-preview/cd99a3cda30d3714...         0.460   0.912   
42   https://p.scdn.co/mp3-preview/15ddd25586b4f62e...         0.233   0.730   
55   https://p.scdn.co/mp3-preview/eae98f3734badfd0...         0.368   0.674   
91   https://p.scdn.co/mp3-preview/af524142f40dcacf...         0.544   0.837   
99   https://p.scdn.co/mp3-preview/fc476336bac928d0...         0.578   0.723   
131  https://p.scdn.co/mp3-preview/3789af5d464b633f...         0.510   0.865   
27   https://p.scdn.co/mp3-preview/1dbf69a32db3b4d2...         0.341   0.867   
114  https://p.scdn.co/mp3-preview/9f9cc354c35bf303...         0.438   0.852   
105  https://p.scdn.co/mp3-preview/f76426030a7cfb44...         0.150   0.637   

     key  loudness  mode  speechiness  acousticness  instrumentalness  \
32     6    -5.491     0       0.0727      0.071100          0.000378   
75     5    -3.444     0       0.0596      0.000579          0.000217   
42    10    -4.855     1       0.0449      0.201000          0.000004   
55     7    -5.859     0       0.0328      0.424000          0.000000   
91     6    -4.618     0       0.0465      0.059600          0.000000   
99     6    -7.949     0       0.0507      0.136000          0.000002   
131    9    -4.793     1       0.0431      0.190000          0.000031   
27     7    -4.727     0       0.0492      0.293000          0.000000   
114    7    -5.567     0       0.0387      0.147000          0.000000   
105   10    -6.177     1       0.0344      0.000843          0.003020   

     liveness  valence    tempo  
32     0.1350    0.476  180.200  
75     0.1040    0.311  134.074  
42     0.1300    0.123  152.972  
55     0.9510    0.124  149.204  
91     0.0698    0.159  128.980  
99     0.1960    0.165  129.054  
131    0.6790    0.451  129.013  
27     0.2570    0.200  182.023  
114    0.9760    0.246   91.061  
105    0.3570    0.174  150.038

One thing it can be noticed is that the song's titles of lyrics data have no " ' ", while the tracks's names from Spotify have. E.g.: song_title: call me when youre sober x track_name: call me when you're sober.

So I'll remove " ' " from all track_name.

In [6]:
df_tracks_evanescence["track_name"] = df_tracks_evanescence["track_name"].apply(lambda x: x.replace("'",""))
df_tracks_within_temptation["track_name"]=df_tracks_within_temptation["track_name"].apply(lambda x: x.replace("'",""))

# Try to merge lyrics data with Spotify's data

In [7]:
df_evanescence_merged = df_lyrics_evanescence.merge(df_tracks_evanescence, left_on='song_title', right_on='track_name')

In [8]:
df_evanescence_merged

song_title  \
0                  lost in paradise   
1                       going under   
2                      lose control   
3                  bring me to life   
4                        cloud nine   
5                         lacrymosa   
6          call me when youre sober   
7                   sweet sacrifice   
8                      imperfection   
9                      farther away   
10                            hello   
11                     say you will   
12                   together again   
13              weight of the world   
14                        your star   
15                        imaginary   
16                          whisper   
17                  everybodys fool   
18                          missing   
19                    never go back   
20                    made of stone   
21                   taking over me   
22                      good enough   
23                       erase this   
24               my heart is broken   
25                      secret door   
26                   the other side   
27                       tourniquet   
28                    what you want   
29                         like you   
30                      my immortal   
31                          lithium   
32                        disappear   
33                   my last breath   
34                 snow white queen   
35                     the only one   
36                             sick   
37                       the change   
38  the last song im wasting on you   
39                    swimming home   
40                          haunted   

                                               lyrics  \
0   I’ve been believing in something so distant, a...   
1   now I will tell you what I've done for you. 50...   
2   You don't remember my name. I don't really car...   
3   how can you see into my eyes. like open doors....   
4   If you want to live, let live.. If you want to...   
5   Out on your own. Cold and alone again. Can thi...   
6   Don't cry to me. If you loved me. You would be...   
7   It's true, we're all a little insane. But it's...   
8   The more you try to fight it. The more you try...   
9   I took their smiles and I made them mine.. I,I...   
10  Playground school bell rings again.. Rain clou...   
11  It's taking you too long to decide. And I don'...   
12  Never thought that I'd be leaving you today. S...   
13  Feels like the weight of the world. Like God i...   
14  I can't see your star. I can't see your star. ...   
15  Paper flowers, paper flowers.... I linger in t...   
16  Catch me as I fall. Say you're here and it's a...   
17  perfect by nature. icons of self indulgence. j...   
18  Please, please forgive me,. But I won't be hom...   
19  Everything is so dark. And I know there's some...   
20  Speak your mind. Like I care. I can see your l...   
21  you don't remember me but I remember you. I li...   
22  Under your spell again. I can't say no to you....   
23  It's too late to change your mind. Even though...   
24  I pulled away to face the pain. I close my eye...   
25  Turn out the lights. Feed the fire till my sou...   
26  Make me whole again. Open your eyes. Taunted b...   
27  I tried to kill the pain. But only brought mor...   
28  Do what you what you want, if you have a dream...   
29  Stay low. Soft, dark, and dreamless. Far benea...   
30  I'm so tired of being here. suppressed by all ...   
31  Lithium, don't want to lock me up inside. Lith...   
32  Hollow, like you don't remember me. Underneath...   
33  Hold on to me, love. You know I can't stay lon...   
34  Stop right, lock the door. Don't look back. Un...   
35  When they all come crashing down- midflight. y...   
36  Embrace the silence. Cause there's nothing tha...   
37  Thought that I was strong. I know the words I ...   
38  Sparkling grey. Through my own veins. Any more...   
39  Way down. I've been way down. Underneath this ...   
40  long lost words whisper slowly to me. still c

In [9]:
df_evanescence_merged.shape[0]/df_lyrics_evanescence.shape[0]

0.47674418604651164

We succeeded in having all information (lyrics, metadatad, and audio features) for 41 songs by simply merging our dataframes. This means that 47.7% of the `Evanescence's songs` for which we retrieved lyrics are represented with additional information.

In [10]:
df_within_temptation_merged = df_lyrics_within_temptation.merge(df_tracks_within_temptation, left_on='song_title', right_on='track_name')
df_within_temptation_merged

song_title  \
0                         lost   
1                the swan song   
2           in perfect harmony   
3                 see who i am   
4                     forsaken   
5                  the howling   
6           what have you done   
7            where is the edge   
8                     forgiven   
9           never ending story   
10           raise your banner   
11                    memories   
12            a dangerous mind   
13                       enter   
14              hand of sorrow   
15                  all i need   
16             pearls of light   
17             stand my ground   
18             our solemn hour   
19            shot in the dark   
20       stairway to the skies   
21                   the dance   
22                     candles   
23     the heart of everything   
24  in the middle of the night   
25              the gatekeeper   
26                 another day   
27                        pale   
28                 the promise   
29           edge of the world   
30           final destination   
31                   the cross   
32           deceiver of fools   
33  the truth beneath the rose   
34                    restless   
35                   somewhere   
36                    aquarius   
37                  why not me   
38                our farewell   
39                      frozen   
40                       caged   
41                      murder   
42                      angels   
43              the last dance   
44                      faster   
45                 deep within   
46               a demons fate   
47                mother earth   
48                   ice queen   
49                  dark wings   
50                fire and ice   
51                       grace   
52                        iron   

                                               lyrics  \
0   My hope is on fire. My dreams are for sale. I ...   
1   Winter has come for me, can't carry on. The ch...   
2   At the end of a closing day. A little child wa...   
3   Is it true what they say,. are we too blind to...   
4   Now the day has come. We are forsaken this tim...   
5   We've been seeing what you want,. You've got u...   
6   What have you done now. I know I'd better stop...   
7   In the shadows it awakes the desire. But you k...   
8   Love you so it hurts my soul. Can you forgive ...   
9   Armies have conquered. And fallen in the end. ...   
10  Blood for freedom. You're pushing me into the ...   
11  In this world you tried. Not leaving me alone ...   
12  Cause something is not right. I follow the sig...   
13  Welcome to my home. The gates of time have ope...   
14  The child without a name grew up to be the han...   
15  I'm dying to catch my breath. Oh why don't I e...   
16  After. The cold darkness,. In the heart of the...   
17  I can see. When you stay low nothing happens. ...   
18  Sanctus Espiritus redeem us from our solemn ho...   
19  I've been left out alone like a damn criminal....   
20  Seven seconds to the rise. Can't believe I'm s...   
21  Someone is dancing on my grave. Sending shiver...   
22  Take away. These hands of darkness. Reaching f...   
23  For the pain and the sorrow. Caused by my mist...   
24  Ooooh, ooooh. Ooooh, ooooh. Ooooh, ooooh. Oooo...   
25  The shadows of the night,. are unleashed again...   
26  I know you are going away. I take my love into...   
27  The world seems not the same. Though I know no...   
28  On behalf of her love. She no longer sleeps. L...   
29  Can I swallow the soil,. Make it all beautiful...   
30  I escaped my final moment. But it's turning ba...   
31  Nothing's ever changed, you still turn away. Y...   
32  He feeds on fear. He feeds on pain. He rules a...   
33  Give me strength to face the truth, the doubt ...   
34  She embraced, with a smile. As she opened the ...   
35  Lost in the darkness. Hoping for a sign. Inste...   
36  I hear your whispers. Break the silence and it...   
37  Some may call

In [11]:
df_within_temptation_merged.shape[0]/df_lyrics_within_temptation.shape[0]

0.7162162162162162

In the case of `Within Temptation`, we got complete information about 71.6% of the songs for which we have lyrics.

# Analysing lyrics

To start I'll build some Wordclouds and analyse some metrics solely based on lyrics since when merging the data not all lyrics were included.

Further investigation probably would increase the lyrics included, but for now we will keep like this.

In this part I will perform the following steps:

1. Clean `lyrics` and save the result in a new column `lyrics_clean`.
2. Apply [TextBlob](https://textblob.readthedocs.io/en/dev/index.html) and create columns with the sentiment property information (`polarity` and `subjectivity`). The polarity score is a float within the range [-1.0, 1.0] where -1.0 is very positive, 0 is neutral, and 1 is very positive. The subjectivity is a float within the range [0.0, 1.0] where 0.0 is very objective and 1.0 is very subjective.
3. Use polarity to define a lyric as positive or negative. This information will be in a binary column `is_positive` where `1` indicates a positive lyric and `0` a negative one.
4. Create feature `lyrics_len` with the lenght of the lyric.
5. Create feature `num_words` with lyric's number of words.
6. Build word clouds
7. Build some graphs.


There are some nice word clouds in R that I wanted to include in this project and in fact I have already something which is included in [GitHub](https://github.com/dpbac/evanescence_and_within_temptation_in_R).

For the R project my intention was to use the original lyrics csv that we have just loaded. However, when developing the R word cloud part some strange characters showed up and although I've tried some R option to try to clean there was no effect. So, I came back to Python and I decided to apply the following line of code:

        unicodedata.normalize('NFKD', text).encode('ascii', 'ignore').decode('utf-8', 'ignore')

found at https://github.com/kaparker/gameofthrones-wordclouds/blob/master/gotwordcloud.py which just fine worked fine.

The following function includes some basic cleaning and also lemmatization. If you need more cleaning the code presented at https://github.com/kaparker/gameofthrones-wordclouds/blob/master/gotwordcloud.py can be very useful. 

In some cases, stemming is used in place of lemmatization. Stemming reduce the words to their root form and because I prefer to have the dictionary form in my clouds I chose to apply lemmatization.

To deal with contraction I've tried both TokTok and Moses and my choice stayed with TokTok, because I didn't like when I got, for instances, gonna as gon na. https://stackoverflow.com/questions/43041039/dont-want-nltk-word-tokenize-to-tokenize-a-single-word-gotta-into-got-and

**TIP**: If you don't succeed in installing pycontractions using `pip install pycontractions` use:

python -m pip install _wheel_file_

in my case _wheel_file_ https://files.pythonhosted.org/packages/a6/f5/d3ec9491c530cbc03af32ca2c6b69b0e89660daeb2856b485d90f9d82e5e/pycontractions-2.0.1-py3-none-any.whl

In [12]:
#https://pypi.org/project/pycontractions/

import unicodedata
from nltk.tokenize import sent_tokenize    
from pycontractions import Contractions
# Specify any model from the gensim.downloader api - that was what worked for me
cont = Contractions(api_key="glove-twitter-100")
# optional, prevents loading on first expand_texts call
cont.load_models()

from nltk.corpus import stopwords
from nltk.tokenize.toktok import ToktokTokenizer
# from nltk.tokenize import word_tokenize
from nltk.stem import WordNetLemmatizer

def nlp_clean(text):
    
    # Removing strange/accented characteres
    text = unicodedata.normalize('NFKD', text).encode('ascii', 'ignore').decode('utf-8', 'ignore')
    
    # Create a list with sentences of the lyrics to be able to apply expand_texts
    sentences = sent_tokenize(text)
    
    
    # Expand words that are contracted
    
    cont.expand_texts(text)
    
    # lower case and keeping what has only of alphanumeric characters - I used toktok because of problems as gonna = gon na
    toktok = ToktokTokenizer()
    tokens = [w for w in toktok.tokenize(text.lower()) if w.isalnum()]
    
    # Removing non stop words in english
    no_stops = [t for t in tokens if t not in stopwords.words('english')]
    
    
    # Build a lemmatized list
    WNlemmatizer = WordNetLemmatizer()
    lem_tokens = [WNlemmatizer.lemmatize(token) for token in no_stops]
   
    return ' '.join(lem_tokens)    

Error initializing LanguageTool


JavaError: can't find Java

In [ ]:
# Cleaning lyrics

df_ly['Review Text'] = preprocess(df['Review Text'])

## Word clouds

In this part I'll be creating some word clouds, simple ones but also using masks. 



1. Performing some cleaning on the text and saving it in a new column
2. Create word clouds 

Clean the text
Create a mask of the image we want to use
Create a word cloud using the mask

# Sentiment Analysis